accepted/2022/simplify-rid-model.md

```diff
@@ -0,0 +1,240 @@
+- Implement a new algorithmic host scheme for RID selection, enabled by default.
+
+The new scheme will be based on the following:
+
+- Each host is built with a hard-coded set of RIDs that it probes for, both triplets and singles.
+- Each host is already built for a specific RID triple, so this change is an evolutionary change.
+- There are processor-agnostic managed-code scenarios where a RID single is relevant, like Windows-only code (for example, accessing the registry) that works the same x64, Arm64 and x86. The same is true for macOS and Linux.
+- The RID that `dotnet --info` returns will always match the first in this hard-coded set of RIDs.
+
+Let's assume that an app is running on Alpine 3.16, the host will probe for the following RIDs:
+
+- `linux-musl-x64`
+- `unix`
+
+Ubuntu would be similar:
+
+- `linux-arm64`
+- `unix`
+
+macOS would be similar:
+
+- `osx-arm64`
+- `unix`
+
+Windows would similar:
+
+- `win-x86`
+- `win`
+
+Note: `win7` and `win10` also exist (as processor-specific and -agnostic). Ideally, we don't have to support those in the host-specific scheme, but we need to do some research on that.
+
+Note: A mix of processor-specific RIDs are used above, purely to demonstrate the range of processor-types supported. Each operating system supports a variety of processor-types, and this is codified in their associated RIDs.
+
+More generally, the host will probe for:
+
+- This environment, RID triplet (OS-CRuntime-arch)
+- This environment, RID single (OS-only).
+
+Note: The abstract `unix` RID is used for macOS and Linux, instead of a concrete RID for those OSes.
+
+The host will implement "first one wins" probing logic for selecting a RID-specific asset.
+
+This behavior only applies for portable apps. RID-specific apps will not probe for RID-specific assets.
+
+There are other RIDs in `runtimes.json` today. This new scheme will not support those. The `runtimes.json` host compat feature will need to be used to support those, which is likely very uncommon. Even if they are quasi-common, we will likely still make this change and require some combination of the ecosystem to adapt and app developers to opt-in to the `runtimes.json` compat mode.
+
+The host and `runtimes.json` must remain compatible. The RIDs known by the host must be a subset of `runtimes.json` RIDs. We may need to update `runtimes.json` to ensure that the RIDs known by the host are present in that file.
+
+
+## Minimum CRT/libc version
+
+This scheme doesn't explicitly define a CRT/libc version. Instead, we will define a minimum CRT/libc version, per major .NET version. That minimum CRT/libc version will form a contract and should be documented, per .NET version. Practically, it defines the oldest OS version that you can run an app on, for a given .NET version.
+
+For .NET 7:
+
+- For Windows, .NET will target the Windows 10 CRT.
+- For Linux with glibc, .NET will target CentOS 7.
```

Comment:
Is that the same for all architectures? For example, in addition to x64, will the CentOS 7 version of glibc also be the baseline for `arm64` and `s390x` architectures?

Reply:
That would be my base assumption.

Reply:
They can be different because they are anyhow incompatible.

If you don't need to support CentOS 7 on an architecture, there may be build time benefits to using something newer.

For s390x, Microsoft isn't building any `linux-s390` artifacts for .NET, and the Red Hat artifacts will use a source-build rid (like `rhel.8-s390x`).
The glibc baseline serves for NuGet package authors that want build binaries that work across a range of distros. The basline should be the lowest version of CentOS/RHEL that supports s390x.

Reply:
I added some text on this.